hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java

```diff
@@ -283,6 +285,8 @@ public HoodieMetadataPayload(Option<GenericRecord> recordOpt) {
             Integer.parseInt(recordIndexRecord.get(RECORD_INDEX_FIELD_FILE_INDEX).toString()),
             Long.parseLong(recordIndexRecord.get(RECORD_INDEX_FIELD_INSTANT_TIME).toString()));
       }
+    } else {
+      this.isDeletedRecord = true;
```

Comment:
in case of `HoodieMetadataPayload#createRecordIndexDelete`, there is discrepancy as the payload is empty record payload. these impl. should be aligned

Reply:
yes. either we should go w/ EmptyHoodieRecordPayload. 
or we should add "IsDelete" flag to recordIndexMetadata similar to how we have added it for other partitions like ColumnStatsMetadata, BloomFilterMetadata, filesystemMetadata

Reply:
@prashantwason I think @xushiyan and @nsivabalan are addressing that many metadata POJOs already take a `isDeleted` flag except the recordIndexMetadata, during the convertion from write status to RLI metadata payload in `HoodieMetadataPayload#createRecordIndexUpdate`, there is a choice is to mark the recordIndexMetadata itself as deleted, which keeps the behavior in line with the other payloads.

But still, in `HoodieMetadataPayload#preCombine`, the merging of RLI infos should be fixed with the `isDeleted` falg just like others.

Reply:
BTW, I don't like how the deletes are handling with existing column stats, even though we already do that does not mean it is correct, maintaining preCombining logic for every meta info is disgusting.

Reply:
I would favor `isDeleted` field in `HoodieRecordIndexInfo` in the schema.

1. It keeps the schema consistent wrt deletes for different MDT index types. Let's say some index types have `isDeleted` and some don't, then it's an added mental burden for developers and also not easy to maintain as we add more indexes. 
2. It gives enough flexibility to have separate delete handling logic for different index types.
3. Let's consider the semantics of the if-else in the `HoodieMetadataPayload` constructor. It is based on different index types. By setting `this.isDeletedRecord = true` in the last else-block we're saying that for all index types other than the ones above, consider the record to be deleted. It does not make much sense from the pov of adding more index types in the future.

Reply:
>> I would favor isDeleted field in HoodieRecordIndexInfo in the schema.
With  this design, for all deletes:
1. We first have to read the existing record from MDT
2. Add an Upsert to the log files
3. Remove the deleted record at time of compaction

The above has performance implication for larger indexes like RI.

With the above design, DELETE from RI are treated exactly as DELETE from dataset - we write a DELETE block to the log file and existing MOR code takes care of it. This is simple.

The reason we could not use this design for MDT is because there is no usecase in MDT where we actually DELETE an entire record. Example:
1. If a file is deleted (during clean), we need to modify the partition_file_list_record to remove that single file

Reply:
The key is not set when the avro data is not set, so I'm wondering whether the `key == null` actually represents it is a delete record.